cmake/SetupCompiler.cmake

```diff
@@ -76,6 +78,8 @@ elseif(CMAKE_COMPILER_IS_GNUCXX)
         set(xad_cxx_extra -mavx2)
     elseif(XAD_SIMD_OPTION STREQUAL AVX512)
         set(xad_cxx_extra  -march=cascadelake)
+    elseif(XAD_SIMD_OPTION STREQUAL NEON)
+        set(xad_cxx_extra  -march=native)
```

Comment:
This is a rather generic flag that compiles for the architecture the compiler is running on, which might break if you're cross-compiling. It's not a flag for NEON SIMD instructions. 

I think we should name the option APPLE_M1 and add the flags `-march=apple-m1` (you can see it [here](https://github.com/llvm/llvm-project/blob/llvmorg-15.0.7/llvm/lib/Target/AArch64/AArch64.td#L1214)). And add another one simply called `NATIVE` with the `-march=native` flag added. Please see if Windows / VC++ also has an equivalent option to `-march=native`.

Just be careful if the compiler is GCC on MacOS - there's no `apple-m1` option in GCC. It's rather [one of these](https://gcc.gnu.org/onlinedocs/gcc/AArch64-Options.html#index-march) - not sure which one would be the M1. 

And note that M1 compatible binaries means the generated code will run on M1 upwards, including M2, M3, ...

Reply:
I originally went with `-march=native` to allow compilation for AppleClang14. I'll add the `NATIVE` option and set that as the default for AppleClang14 until I find the appropriate flag. 

Thank you for your feedback, I'll take your feedback and make the appropriate changes over the next week or two.

Reply:
I've just made a commit. If you look at the workflows, you'll see the AppleClang15 build is failing due to the flag. From my research, the macoslatest on GitHub Actions should support this, but I'll look into it further.

Unrelated to the above, but the `Mac (debug)` build is failing on `coverage baseline`. I seen both the Mac debug and release builds failed as of 3 weeks ago on the original repo, so it's possible this coverage baseline failure was an issue down the road already. It would be handy to fix this in my branch since I've already fixed `Mac (release).` Any idea on where to start on this?

Reply:
There's probably a difference between the `AppleClang` official compiler shipped with XCode, and the Clang 15 compiler installed separately. I think you may have to research which ARM architecture the M1 is. From what I can see it's the Apple A14, which is a [ARMv8.5-A architecture](https://en.wikipedia.org/wiki/Apple_A14). The [GCC flag would be](https://gcc.gnu.org/onlinedocs/gcc/AArch64-Options.html#index-march) `-march=armv8.5-a`. And I believe the same flag would work with Clang as well (incl. non `AppleClang`). This might also work with Clang 14 (very likley).

Re coverage baseline - I am not sure why these failures happen and it requires some research. It my be compatibility issues with between `lcov` and the coverage annotations created by Clang. There might also be a totally different way to capture coverage on Mac with Clang. Perhaps you can give that a try on your box. The goal is to generate a coverage `lcov.info` file, which has information about the lines covered in tests. This is uploaded to coveralls for statistics. We run coverage on both Mac and Linux and join it since there are many parts in the code that are platform-specific (ifdef apple, etc.). Collecting on Linux only would not give a good picture.

Reply:
Would it be appropriate to set the default flag for `APPLE_M1` to `-march=armv8.5-a` then? It should cover all scenarios.

As for coverage baseline, I'll do my research.

Reply:
Some time has passed since my last commit and during this I've updated the Mac build to use a wider matrix; split into `macos-13` & `macos-latest` and `Debug` & `Release` for each.

As for the baseline coverage issue, I've done some research and I've seen people with similar issues have discovered it's a versioning issue, since new versions of lcov have compatibility issues with gcov. I'm not convinced that's the issue in our case. The `Linux (Debug)` build uses `lcov 1.14-2` while `Mac (Debug)` uses `lcov 2.1`.  I'll make the Mac build use a downgraded version of lcov and see if that fixes the issue.

After that, there's one final issue with the build and it's the `canEvaluateLongExpressionsLikeHestonAdjoint` test case and its forward mode counterpart. As per the ctest verbose output (both tests have the same output):
```
Expected equality of these values:
    value(y)
        Which is: 4.1619803753616083
    yd
        Which is: 4.1619803753614786
```
You'll find the inequality starts at 1e-12, which is around where decimal uncertainty arises for computers generally - I'm not sure if we can achieve equality at the 1e-16th place with any sort of software change. Unless I'm missing something, these two tests fail due to a universal hardware limitation (error propagation of binary division) and so I suggest allowing a tolerance level for this test.

Reply:
Yes, please put a tolerance level - see other tests that do that already.